### PR TITLE
Add multidex flag and automatic multidex support

### LINKED ISSUE #21009: Flutter Gradle should automatically add MultiDex
## Steps to Reproduce

Build app with more than 64K methods.

## Recommended Changes
1. Project template or `flutter.gradle` script should set the `multiDexEnabled` property to true, and add the `multiDexKeepFile file('multidex-flutter.txt')` config file. Example: [multidex-flutter.txt](https://github.com/flutter/flutter/files/2320127/multidex-flutter.txt)
    ```txt
    io/flutter/app/FlutterApplication.class
    io/flutter/view/FlutterMain.class
    io/flutter/util/PathUtils.class
    ```

2. Project template or `flutter.gradle` script should add the multidex dependency.
    `implementation 'com.android.support:multidex:1.0.3'`

3. FlutterApplication should have the following addition.
    ```java
    @Override
    protected void attachBaseContext(Context base)
    {
        super.attachBaseContext(base);
        MultiDex.install(this);
    }
    ```

**Adding the `multiDexKeepFile` is important as the `FlutterApplication` class and its dependencies are required to me in the primary dex file or the Application class may its onCreate lifecycle event.**

```
Flutter 0.5.7 • channel v0.5.7 • https://github.com/slightfoot/flutter
Framework • revision 66091f9696 (7 weeks ago) • 2018-07-09 12:52:41 -0700
Engine • revision 6fe748490d
Tools • Dart 2.0.0-dev.63.0.flutter-4c9689c1d2
```

### PR DESCRIPTION
**Pull request by GaryQian**
This PR adds multi dex support to the flutter_tool. Resolves https://github.com/flutter/flutter/issues/21009

When the multi dex gradle build error is detected, Flutter will prompt the developer whether they wish to enable multi dex support. If yes, android/app/src/main/java/io/flutter/app/FlutterMultiDexApplication.java will be added and the build will be retried.

Once enabled, the existence of the files will be used to determine if the multidex gradle groovy configs and dependencies should added to the app. In practice, this means that apps will build without issue after selecting enable.

Developers who don't want this may pass the --no-multidex flag to disable this automated support for flutter run and flutter build commands.

FlutterMultiDexApplication is codegen-ed rather than included in the engine to reduce engine dependencies and avoid the overhead to non-multidex apps.

### CODE REVIEW

**Comment by GaryQian** [comment]
Would prefer if the approach is reviewed first, will add tests shortly.

File: packages/flutter_tools/gradle/flutter.gradle [hunk 1]
```diff
@@ -149,6 +149,24 @@ class FlutterPlugin implements Plugin<Project> {
...
+        if (project.hasProperty("multi-dex-enabled") &&
+            project.property("multi-dex-enabled").toBoolean() &&
+            project.android.defaultConfig.minSdkVersion <= 20) {
+            project.android {
+                defaultConfig {
+                    multiDexEnabled true
+                }
+                buildTypes {
+                    release {
+                        multiDexKeepFile project.file('multidex-flutter-keepfile.txt')
```
Comment (blasten) [inline packages/flutter_tools/gradle/flutter.gradle:161]: could this file be moved to the flutter installation? 

flutter_proguard_rules.pro had a similar issue:
https://github.com/flutter/flutter/blob/e7a9adeeabccf99d67321dc7a87fdf3820f06912/packages/flutter_tools/gradle/flutter.gradle#L188

File: packages/flutter_tools/lib/src/runner/flutter_command.dart [hunk 2]
```diff
@@ -815,6 +815,16 @@ abstract class FlutterCommand extends Command<void> {
     );
   }
 
+  void addMultiDexOption({ bool hide = false }) {
+    argParser.addFlag('multi-dex',
```
Comment (blasten) [inline packages/flutter_tools/lib/src/runner/flutter_command.dart:819]: as weird as it sounds, most docs I found refer to "multidex" as a single word.

**Comment by blasten** [comment]
other than the two comments, approach LGTM

File: packages/flutter_tools/lib/src/runner/flutter_command.dart (hunk 2 above)
Comment (GaryQian) [reply packages/flutter_tools/lib/src/runner/flutter_command.dart:819]: I was debating to use the one word vs two word version. I think technically, it is two, but I did notice that a lot of docs used the single word, so I will probably just switch to the single word to remain consistent.

File: packages/flutter_tools/gradle/flutter.gradle (hunk 1 above)
Comment (GaryQian) [reply packages/flutter_tools/gradle/flutter.gradle:161]: Yeah!

**Comment by GaryQian** [comment]
In testing with lower API devices, turns out multidex is incompatible with reflection. Moving forwards with an alternative solution.

**Comment by GaryQian** [comment]
New approach:

- codegen FlutterMultiDexApplication.java, thus avoiding adding overhead to non-multidex apps by not including it as an engine dependency. This also makes it easier for developers to integrate it with any custom android implementations.
- Use manifest variable injection to automatically inject either android.app.Application or io.flutter.app.FlutterMultiDexApplication depending on if multidex is enabled or not. The template should now include a variable for `<application android:name="{$applicationName}>` property in AndroidManifest.xml
- For older apps that do not have the variable injection in their manifest, we ask them to manually add the line.
- Newly generated apps should already have the variable injection and can automatically enable multidex.

**Review by blasten — APPROVED** [review]
This is awesome!

**Pull request merged**